Engine.UnitTests/ExpressionTest.cs

```diff
@@ -0,0 +1,282 @@
+            ExpressionManager.Update(step);
+            Assert.AreEqual(1.0 + 2.0, step.DelaySecs);
+            step.DelaySecs = 0.0;
+            var plan = new TestPlan();
+            plan.ChildTestSteps.Add(step);
+
+            var xml = plan.SerializeToString();
+
+            plan = (TestPlan)new TapSerializer().DeserializeFromString(xml);
+            step = (DelayStep)plan.ChildTestSteps[0];
+            var expr2 = ExpressionManager.GetExpression(step, TypeData.GetTypeData(step).GetMember(nameof(step.DelaySecs)));
+            Assert.IsFalse(string.IsNullOrWhiteSpace(expr2));
+            ExpressionManager.Update(step);
+            Assert.AreEqual(1.0 + 2.0, step.DelaySecs);
+        }
+        
+        [Test]
+        public void StepWithExpression2Test()
+        {
+            var step = new LogStep();
+            ExpressionManager.SetExpression(step, TypeData.GetTypeData(step).GetMember(nameof(step.LogMessage)), "The result is: {1.0 + 2.0}");
+            ExpressionManager.Update(step);
+            Assert.AreEqual("The result is: 3", step.LogMessage);
+        }
+
+        [TestCase("The number is {1 + 2}.", "The number is 3.")]
+        [TestCase("{1 - 2}. {4}. {10 /5}.", "-1. 4. 2.")]
+        [TestCase("", "")]
+        [TestCase("123abc", "123abc")]
+        [TestCase("{{123abc}}{5 + 10}{{123}}", "{123abc}15{123}")]
+        [TestCase("{5}", "5")]
+        [TestCase("{5+4+1}", "10")]
+        [TestCase("{-5+-4+-1}", "-10")]
+        [TestCase("{-5+4+1}", "0")]
+        [TestCase("a b {$\"c {1 + 5}\"} e f g", "a b c 6 e f g")]
+        
+        [TestCase("{floor(3.5)}", "3")]
+        [TestCase("{floor(1 + 3.5)}", "4")]
+        
+        [TestCase("{ceiling(3.5 + 0.6)}", "5")]
+        [TestCase("{ceiling(3.5 + 0.6) + floor(-1.5)}", "3")]
+        [TestCase("{round(1.11111111, 2)}", "1.11")]
+        [TestCase("{sin(π * 2.0)} {cos(π * 2.0)}", "0 1")]
+        [TestCase("{cos(π * 2.0)} {sin(π * 2.0)}", "1 0")]
+        [TestCase("{π}", "3.14159265358979")]
+        [TestCase("{true}", "True")]
+        [TestCase("{false}", "False")]
+        [TestCase("{abs(-1.0)}", "1")]
+        [TestCase("{empty(\"\")}", "True")]
+        [TestCase("{empty(\"asd\")}", "False")]
+        public void StringExpressionBasicTest(string expression, string expectedResult)
+        {
+            var builder = new ExpressionCodeBuilder();
+            var ast0 = builder.ParseStringInterpolation(expression);
+            var ast = ast0.Unwrap();
+            var lmb = builder.GenerateLambda(ast, ParameterData.Empty, typeof(string));
+            var result = lmb.Unwrap().DynamicInvoke();
+            Assert.AreEqual(expectedResult, result);
+        }
+
+        [TestCase( "123 {floor(1.5} 321432", "Unexpected symbol '}'.", null)]
+        [TestCase( "123 {1.5) 321432", "Unexpected symbol ')'.", null)]
+        public void StringExpressionParseErrors(string errorExpression, string parseError, string compileError)
```

Comment:
If both of these have null for the compile error, is there a reason its included as a parameter?